sockeye/inference.py

```diff
@@ -1089,22 +1108,55 @@ def _beam_search(self,
             attentions[:, t, :] = attention_scores
             lengths += mx.nd.cast(1 - mx.nd.expand_dims(finished, axis=1), dtype='float32')
 
-            # (6) determine which hypotheses in the beam are now finished
+
+            # (6) save beam history
+            # TODO hussed: check speed, if slow,  condition this block on some CLI flag so it doesn't run in prod
```

Comment:
I think this should be off by default to reduce overhead, for example lets implement another output_handler that enables this feature, otherwise we leave it off.
I would also factor this out into some kind of `_update_beam_history()` function to make the length of _beam_search smaller.

Reply:
I agree, the problem I had was how to implement the switch. Where should I get the condition the `if store_beam:` should check? Should it just be an extra arg to `Translator().__init__()`? It would have to pass down quite a few functions to reach here.

About factoring out, I would rather leave it here for now, it's not too much code. We had discussed with Matt to move the beam search function to a more flexible and readable class. I'll probably work on that after the visualization is done.

Reply:
I think making this a member for `Translator` is reasonable, just like `restrict_lexicon`. Then you can directly access it it the beam search function, no?

something like:
```
if output-handler-type == `beam_visualization`:
  output_handler = <implementation to handle dumping histories to json>

translator = Translator(..., store_beam=output-handler-type == `beam_visualization`,...)

Reply:
Sounds good, will do that then!